From a20e5174eeb2c1133edbfbb25c8db75bf1825295 Mon Sep 17 00:00:00 2001
From: Dezhi Kong <dezhi.kong@amlogic.com>
Date: Wed, 12 Feb 2020 14:16:56 +0800
Subject: [PATCH 7/7] vpp: add nv12 format support [1/1]

PD#SWPL-20468

Problem:
nv12 format is not support in vpp

Solution:
add nv12 format support

Verify:
on U212

Change-Id: I1e621bede63a283931743c2cd3a5ea1548149368
Signed-off-by: Dezhi Kong <dezhi.kong@amlogic.com>
---
 drivers/amlogic/media/video_sink/video_hw.c | 18 ++++++++++++++----
 include/linux/amlogic/media/vfm/vframe.h    |  1 +
 2 files changed, 15 insertions(+), 4 deletions(-)

diff --git a/drivers/amlogic/media/video_sink/video_hw.c b/drivers/amlogic/media/video_sink/video_hw.c
index 874ca6c26bad..2ac18b914b10 100644
--- a/drivers/amlogic/media/video_sink/video_hw.c
+++ b/drivers/amlogic/media/video_sink/video_hw.c
@@ -833,11 +833,15 @@ static void vd1_set_dcu(
 	if (type & VIDTYPE_VIU_NV21)
 		VSYNC_WR_MPEG_REG_BITS(
 			VD1_IF0_GEN_REG2 +
-			vd_off, 1, 0, 1);
+			vd_off, 1, 0, 2);
+	else if (type & VIDTYPE_VIU_NV12)
+		VSYNC_WR_MPEG_REG_BITS(
+			VD1_IF0_GEN_REG2 +
+			vd_off, 2, 0, 2);
 	else
 		VSYNC_WR_MPEG_REG_BITS(
 			VD1_IF0_GEN_REG2 +
-			vd_off, 0, 0, 1);
+			vd_off, 0, 0, 2);
 
 	/* FIXME: don't use glayer_info[0].reverse */
 	if (glayer_info[0].reverse) {
@@ -870,6 +874,7 @@ static void vd1_set_dcu(
 		if (is_crop_left_odd(frame_par)) {
 			if (!(type & VIDTYPE_PRE_INTERLACE) &&
 			    ((type & VIDTYPE_VIU_NV21) ||
+			     (type & VIDTYPE_VIU_NV12) ||
 			     (type & VIDTYPE_VIU_422)))
 				hphase = 0x8 << HFORMATTER_PHASE_BIT;
 		}
@@ -1240,11 +1245,15 @@ static void vd2_set_dcu(
 	if (type & VIDTYPE_VIU_NV21)
 		VSYNC_WR_MPEG_REG_BITS(
 			VD2_IF0_GEN_REG2 +
-			vd_off, 1, 0, 1);
+			vd_off, 1, 0, 2);
+	else if (type & VIDTYPE_VIU_NV12)
+		VSYNC_WR_MPEG_REG_BITS(
+			VD2_IF0_GEN_REG2 +
+			vd_off, 2, 0, 2);
 	else
 		VSYNC_WR_MPEG_REG_BITS(
 			VD2_IF0_GEN_REG2 +
-			vd_off, 0, 0, 1);
+			vd_off, 0, 0, 2);
 
 	/* FIXME: don't use glayer_info[1].reverse */
 	if (glayer_info[1].reverse)
@@ -1269,6 +1278,7 @@ static void vd2_set_dcu(
 		if (is_crop_left_odd(frame_par)) {
 			if (!(type & VIDTYPE_PRE_INTERLACE) &&
 			    ((type & VIDTYPE_VIU_NV21) ||
+			     (type & VIDTYPE_VIU_NV12) ||
 			     (type & VIDTYPE_VIU_422)))
 				hphase = 0x8 << HFORMATTER_PHASE_BIT;
 		}
diff --git a/include/linux/amlogic/media/vfm/vframe.h b/include/linux/amlogic/media/vfm/vframe.h
index 8af0eff3a147..49a78c62c345 100644
--- a/include/linux/amlogic/media/vfm/vframe.h
+++ b/include/linux/amlogic/media/vfm/vframe.h
@@ -32,6 +32,7 @@
 #define VIDTYPE_INTERLACE_FIRST         0x8
 #define VIDTYPE_MVC                     0x10
 #define VIDTYPE_NO_VIDEO_ENABLE         0x20
+#define VIDTYPE_VIU_NV12                0x80
 #define VIDTYPE_VIU_422                 0x800
 #define VIDTYPE_VIU_FIELD               0x1000
 #define VIDTYPE_VIU_SINGLE_PLANE        0x2000
-- 
2.24.1

